rotkehlchen/chain/evm/decoding/extrafi/cache.py

```diff
@@ -0,0 +1,115 @@
+    from rotkehlchen.user_messages import MessagesAggregator
+
+
+def chunk_range(iterable: Sequence, size: int) -> Iterator:
+    """Split into chunks of size `size` the iterable provided"""
+    it = iter(iterable)
+    return iter(lambda: tuple(islice(it, size)), ())
+
+
+def get_existing_reward_pools(chain_id: ChainID) -> tuple[set[ChecksumEvmAddress]]:
+    """Get extrafi rewards pools for extrafi lending"""
+    with GlobalDBHandler().conn.read_ctx() as cursor:
+        return ({
+            string_to_evm_address(address) for address in globaldb_get_general_cache_values(
+                cursor=cursor,
+                key_parts=(CacheType.EXTRAFI_REWARD_CONTRACTS, str(chain_id.serialize())),
+            )
+        },)
+
+
+def query_extrafi_data(
+        inquirer: 'EvmNodeInquirer',
+        cache_type: Literal[CacheType.EXTRAFI_NEXT_RESERVE_ID],
+        msg_aggregator: 'MessagesAggregator',
+) -> list[ChecksumEvmAddress] | None:
+    """Query and store information of rewards pools
+    Updates the last queried ts when executed.
+    """
+    lending_contract = EvmContract(
+        address=EXTRAFI_POOL_CONTRACT,
+        abi=inquirer.contracts.abi('EXTRAFI_LENDING'),
+        deployed_block=96265067,
+    )
+    cache_key = (cache_type, str(inquirer.chain_id.serialize_for_db()))
+    next_reserve_id = lending_contract.call(
+        node_inquirer=inquirer,
+        method_name='nextReserveId',
+    )
+
+    with GlobalDBHandler().conn.read_ctx() as cursor:
+        saved_next_reserve_id = globaldb_get_unique_cache_value(
+            cursor=cursor,
+            key_parts=cache_key,
+        )
+        if saved_next_reserve_id and str(next_reserve_id) == saved_next_reserve_id:
+            with GlobalDBHandler().conn.write_ctx() as write_cursor:
+                globaldb_set_unique_cache_value_at_ts(
+                    write_cursor=write_cursor,
+                    key_parts=cache_key,
+                    value=saved_next_reserve_id,
+                    timestamp=ts_now(),
+                )
+                return None
+
+        reward_addresses = []
+        for chunk in chunk_range(range(1, next_reserve_id), 20):
+            calls = [(
+                EXTRAFI_POOL_CONTRACT,
+                lending_contract.encode(method_name='getStakingAddress', arguments=[idx]),
+            ) for idx in chunk]
+
+            try:
+                outputs = inquirer.multicall_2(require_success=False, calls=calls)
+            except RemoteError:
+                outputs = [(False, b'') for _ in chunk]
+
+            for idx, output in enumerate(outputs):
+                status, encoded_data = output
+                if status is False:
+                    continue
+
+                reward_addresses.append(to_checksum_address(lending_contract.decode(
+                    result=encoded_data,
+                    method_name='getStakingAddress',
+                    arguments=[chunk[idx]],
```

Comment:
What if idx is out of range? Can that happen if any contract returns something wrong/weird or a remote call fails? If yes you need to guard for it, log an error and continue

Reply:
It can't happen that multicall_2 returns False when the query fails and always the length of the output is the same as in the input. the result is tuples of true/false for the status and the actual output of the call